os/drivers/pm/pm.c

```diff
@@ -0,0 +1,204 @@
+            		g_pm_timer.timer_type = PM_NO_TIMER;
+        	}
+        	ret = OK;
+        	break;
+        /****************************************************************************
+         * Description:
+         *  Unlock PM from PM_NORMAL state by invoking pm_relax(). 
+         ****************************************************************************/
+        case PMIOC_UNLOCK:
+            	pm_relax(PM_IDLE_DOMAIN, PM_NORMAL);
+            	fvdbg("State unlocked!\n");
+            	g_pm_timer.timer_type = PM_NO_TIMER;
+            	ret = OK;
+            	break;
+        /****************************************************************************
+         * Description:
+         *  Make the board wake up after given time (arg) when board sleeps.
+         ****************************************************************************/
+        case PMIOC_WAKEUP_TIMER:
+        	if (g_pm_timer.timer_type == PM_NO_TIMER && arg > 0) {
+        		pm_set_timer(PM_WAKEUP_TIMER, arg);
+        	}	 
+        	ret = OK;
+        	break;
+        /****************************************************************************
+         * Description:
+         *  Sets timer interval for next sleep cycle after sending of wifi keep 
+         *  alive signal from UART. It also stores the current time tick 
+         *  in g_pm_timer.last_wifi_alive_send_time that will be used to calculate the 
+         *  duration of required sleep later just before sleep. If another
+         *  interrupt occurs during this interval, the system will still
+         *  wake up.
+         *  Also we dont want to do pm_set_timer to PM_WAKEUP_TIMER if the
+         *  system is already in PM_LOCK_TIMER state. If system is in 
+         *  PM_LOCK_TIMER state, then power_unlock_write can only change that state.
+         *
+         ****************************************************************************/
+        case PMIOC_SET_WIFI_WAKEUP:
+        	g_pm_timer.last_wifi_alive_send_time = clock_systimer();
+        	if (g_pm_timer.timer_type == PM_NO_TIMER && arg > 0) {
+        		pm_set_timer(PM_WAKEUP_TIMER, arg);
+        	}
+        	ret = OK;
+        	break;
+        /****************************************************************************
+         * Description:
+         *   apps can use this method to get the current time tick.
+         *
+         * Returned Value:
+         *  Time tick value of current time
+         ****************************************************************************/
+        case PMIOC_TIMETICK:
+        	*((size_t *)arg) = clock_systimer();
+			ret = OK;
+        	break;
+
+#ifdef CONFIG_PM_DVFS
+        /****************************************************************************
+        * Description:
+        *  Change the operating frequency of the core to save power
+        *  consumption during active mode.
+        ****************************************************************************/
```

Comment:
Good for explanation of each command. But it should be in the header, not the source code. Because user will check the header, not the source file.